libzerocash/ZerocashParams.cpp

```diff
@@ -18,130 +18,148 @@
-        std::stringstream ssVerification;
-        std::ifstream fileVerification(pathToVerificationParams, std::ios::binary);
+    ssVerification << fileVerification.rdbuf();
+    fileVerification.close();
 
-        if(!fileVerification.is_open()) {
-            throw ZerocashException("Could not open verification key file.");
-        }
+    ssVerification.rdbuf()->pubseekpos(0, std::ios_base::in);
 
-        ssVerification << fileVerification.rdbuf();
-        fileVerification.close();
+    r1cs_ppzksnark_verification_key<ZerocashParams::zerocash_pp> vk_temp;
+    ssVerification >> vk_temp;
 
-        ssVerification.rdbuf()->pubseekpos(0, std::ios_base::in);
-
-        r1cs_ppzksnark_verification_key<ZerocashParams::zerocash_pp> vk_temp2;
-        ssVerification >> vk_temp2;
+    return zerocash_pour_verification_key<ZerocashParams::zerocash_pp>(
+        libzerocash::ZerocashParams::numPourInputs,
+        libzerocash::ZerocashParams::numPourOutputs,
+        std::move(vk_temp)
+    );
+}
 
-        params_vk_v1 = new zerocash_pour_verification_key<ZerocashParams::zerocash_pp>(this->numPourInputs,
-                                                                                       this->numPourOutputs,
-                                                                                       std::move(vk_temp2));
-    }
-    else {
-        params_vk_v1 = NULL;
-    }
+ZerocashParams::ZerocashParams(
+    const unsigned int tree_depth,
+    zerocash_pour_keypair<ZerocashParams::zerocash_pp> *keypair
+) :
+    treeDepth(tree_depth)
+{
+    params_pk_v1 = new zerocash_pour_proving_key<ZerocashParams::zerocash_pp>(keypair->pk);
+    params_vk_v1 = new zerocash_pour_verification_key<ZerocashParams::zerocash_pp>(keypair->vk);
 }
 
-void ZerocashParams::initV1Params()
+ZerocashParams::ZerocashParams(
+    const unsigned int tree_depth,
+    zerocash_pour_proving_key<ZerocashParams::zerocash_pp>* p_pk_1,
+    zerocash_pour_verification_key<ZerocashParams::zerocash_pp>* p_vk_1
+) :
+    treeDepth(tree_depth)
 {
-    ZerocashParams::zerocash_pp::init_public_params();
-    kp_v1 = new zerocash_pour_keypair<ZerocashParams::zerocash_pp>(zerocash_pour_ppzksnark_generator<ZerocashParams::zerocash_pp>(this->numPourInputs,
-                                                                                                                                  this->numPourOutputs,
-                                                                                                                                  this->treeDepth));
+    if (p_pk_1 == NULL) {
+        params_pk_v1 = NULL;
+    } else {
+        params_pk_v1 = new zerocash_pour_proving_key<ZerocashParams::zerocash_pp>(*p_pk_1);
+    }
 
-	params_pk_v1 = &kp_v1->pk;
-	params_vk_v1 = &kp_v1->vk;
+    if (p_vk_1 == NULL) {
+        params_vk_v1 = NULL;
+    } else {
+        params_vk_v1 = new zerocash_pour_verification_key<ZerocashParams::zerocash_pp>(*p_vk_1);
+    }
```

Comment:
[BLOCK] Let's add a precondition assertion that ``(p_pk_1 != NULL) || (p_vk_1 != NULL)``. I believe an assertion is correct because the caller *must* be responsible for meeting this condition.